File: backend/neural_mri/schemas/scan.py
from __future__ import annotations

from pydantic import BaseModel


# --- T1: Structural Scan ---


class StructuralScanRequest(BaseModel):
    pass


class LayerStructure(BaseModel):
    layer_id: str  # "embed", "blocks.0.attn", "blocks.0.mlp", "unembed"
    layer_type: str  # "embedding" | "attention" | "mlp" | "output"
    layer_index: int | None = None
    param_count: int
    shape_info: dict


class ConnectionInfo(BaseModel):
    from_id: str
    to_id: str
    type: str  # "sequential" | "residual"


class StructuralData(BaseModel):
    model_id: str
    scan_mode: str = "T1"
    total_params: int
    layers: list[LayerStructure]
    connections: list[ConnectionInfo]
    metadata: dict


# --- T2: Weight Scan ---


class WeightScanRequest(BaseModel):
    layers: list[str] | None = None  # None = all layers


class LayerWeightStats(BaseModel):
    layer_id: str
    component: str
    mean: float
    std: float
    min_val: float
    max_val: float
    l2_norm: float
    shape: list[int]
    num_outliers: int
    histogram: list[float]  # 20-bin histogram


class WeightData(BaseModel):
    model_id: str
    scan_mode: str = "T2"
    layers: list[LayerWeightStats]
    metadata: dict


# --- fMRI: Activation Scan ---


class ActivationScanRequest(BaseModel):
    prompt: str
    layers: list[str] | None = None  # None = all layers
    aggregation: str = "l2"  # "l2" | "mean"


class LayerActivation(BaseModel):
    layer_id: str
    activations: list[float]  # per-token scalar activation (0-1 normalized)
    per_head: list[list[float]] | None = None  # [n_heads][n_tokens] for attn layers


class ActivationData(BaseModel):
    model_id: str
    scan_mode: str = "fMRI"
    tokens: list[str]  # tokenized prompt strings
    layers: list[LayerActivation]
    metadata: dict


# --- DTI: Circuit Scan ---


class CircuitScanRequest(BaseModel):
    prompt: str
    target_token_idx: int = -1  # which output token to trace (-1 = last)


class PathwayConnection(BaseModel):
    from_id: str
    to_id: str
    strength: float  # 0-1 normalized importance
    is_pathway: bool  # True if strength > threshold


class AttentionHead(BaseModel):
    layer_idx: int
    head_idx: int
    pattern: list[list[float]]  # [seq_len, seq_len] attention weights


class ComponentImportance(BaseModel):
    layer_id: str
    importance: float  # 0-1 normalized
    is_pathway: bool


class CircuitData(BaseModel):
    model_id: str
    scan_mode: str = "DTI"
    tokens: list[str]
    target_token_idx: int
    connections: list[PathwayConnection]
    components: list[ComponentImportance]
    attention_heads: list[AttentionHead]
    metadata: dict


# --- FLAIR: Anomaly Scan ---


class AnomalyScanRequest(BaseModel):
    prompt: str


class LayerAnomaly(BaseModel):
    layer_id: str
    anomaly_scores: list[float]  # per-token anomaly score (0-1)
    kl_scores: list[float]  # per-token KL divergence (0-1 normalized)
    entropy_scores: list[float]  # per-token entropy (0-1 normalized)


class AnomalyData(BaseModel):
    model_id: str
    scan_mode: str = "FLAIR"
    tokens: list[str]
    layers: list[LayerAnomaly]
    metadata: dict

File: frontend/src/types/scan.ts
export interface LayerStructure {
  layer_id: string;
  layer_type: 'embedding' | 'attention' | 'mlp' | 'output';
  layer_index: number | null;
  param_count: number;
  shape_info: Record<string, number>;
}

export interface ConnectionInfo {
  from_id: string;
  to_id: string;
  type: string;
}

export interface StructuralData {
  model_id: string;
  scan_mode: 'T1';
  total_params: number;
  layers: LayerStructure[];
  connections: ConnectionInfo[];
  metadata: Record<string, unknown>;
}

export interface LayerWeightStats {
  layer_id: string;
  component: string;
  mean: number;
  std: number;
  min_val: number;
  max_val: number;
  l2_norm: number;
  shape: number[];
  num_outliers: number;
  histogram: number[];
}

export interface WeightData {
  model_id: string;
  scan_mode: 'T2';
  layers: LayerWeightStats[];
  metadata: Record<string, unknown>;
}

// --- fMRI: Activation Scan ---

export interface LayerActivation {
  layer_id: string;
  activations: number[]; // per-token scalar (0-1 normalized)
  per_head?: number[][] | null; // [n_heads][n_tokens] for attn layers
}

export interface ActivationData {
  model_id: string;
  scan_mode: 'fMRI';
  tokens: string[];
  layers: LayerActivation[];
  metadata: Record<string, unknown>;
}

// --- DTI: Circuit Scan ---

export interface PathwayConnection {
  from_id: string;
  to_id: string;
  strength: number;
  is_pathway: boolean;
}

export interface ComponentImportance {
  layer_id: string;
  importance: number;
  is_pathway: boolean;
}

export interface AttentionHead {
  layer_idx: number;
  head_idx: number;
  pattern: number[][];
}

export interface CircuitData {
  model_id: string;
  scan_mode: 'DTI';
  tokens: string[];
  target_token_idx: number;
  connections: PathwayConnection[];
  components: ComponentImportance[];
  attention_heads: AttentionHead[];
  metadata: Record<string, unknown>;
}

// --- FLAIR: Anomaly Scan ---

export interface TokenPredictionLens {
  token: string;
  prob: number;
}

export interface LayerAnomaly {
  layer_id: string;
  anomaly_scores: number[];
  kl_scores: number[];
  entropy_scores: number[];
  top_predictions?: TokenPredictionLens[][] | null; // [seq_len][top_k]
}

export interface AnomalyData {
  model_id: string;
  scan_mode: 'FLAIR';
  tokens: string[];
  layers: LayerAnomaly[];
  metadata: Record<string, unknown>;
}

// --- SAE: Sparse Autoencoder Feature Scan ---

export interface SAEFeatureInfo {
  feature_idx: number;
  activation: number;
  activation_normalized: number;
  neuronpedia_url: string | null;
}

export interface SAETokenFeatures {
  token_idx: number;
  token_str: string;
  top_features: SAEFeatureInfo[];
}

export interface SAEData {
  model_id: string;
  scan_mode: 'SAE';
  prompt: string;
  layer_idx: number;
  hook_name: string;
  d_sae: number;
  tokens: string[];
  token_features: SAETokenFeatures[];
  reconstruction_loss: number;
  sparsity: number;
  heatmap_feature_indices: number[];
  heatmap_values: number[][];
  metadata: Record<string, unknown>;
}

export interface SAEInfoResponse {
  available: boolean;
  model_id: string | null;
  release?: string;
  layers: number[];
  d_sae: number;
  has_neuronpedia?: boolean;
}

// --- WebSocket Messages ---

export type WsMessageIn =
  | { type: 'scan_stream'; mode: 'fMRI' | 'DTI'; prompt: string }
  | { type: 'ping' };

export type WsMessageOut =
  | { type: 'info'; message: string }
  | { type: 'scan_start'; mode: string; tokens: string[]; n_layers: number; seq_len: number }
  | { type: 'activation_frame'; token_idx: number; layers: { layer_id: string; activation: number }[] }
  | { type: 'attention_pattern'; layer_idx: number; head_idx: number; pattern: number[][] }
  | { type: 'component_importance'; layer_id: string; importance: number; is_pathway: boolean }
  | { type: 'scan_complete'; compute_time_ms: number }
  | { type: 'error'; message: string }
  | { type: 'pong' };
